### PR TITLE
Text rendering

### PR DESCRIPTION
**Pull request by kyle-sammons**
Shitty text rendering, yo. Accidentally committed the .DS_Stores.

### CODE REVIEW

**Comment by kyle-sammons** [comment]
Source for 99% of my code; https://github.com/mankyKitty/sdl2-ttf/blob/master/examples/font_test.hs

File: app/Main.hs [hunk 1]
```diff
@@ -26,12 +28,35 @@ initScreen width height = do
         SDL.setWindowTitle window s
     return $ SDLContext window renderer
 
+getWindow :: SDLContext -> SDL.Window
+getWindow (SDLContext window _) = window
+
+getRenderer :: SDLContext -> SDL.Renderer
+getRenderer (SDLContext _ renderer) = renderer
```
Comment (forestbelton) [inline app/Main.hs]: `contextWindow` and `contextRenderer` are exactly these two functions

File: app/Main.hs [hunk 2]
```diff
@@ -26,12 +28,35 @@ initScreen width height = do
...
+    TTF.withInit $ do
+        font <- TTF.openFont arial 150
+        textSurface <- TTF.renderUTF8Solid font "test" (SDL.Color 255 255 255 0)
+        textTexture <- SDL.createTextureFromSurface (getRenderer ctx) textSurface
+        loop ptrEvent ctx textTexture 
+
+loop :: Ptr SDL.Event -> SDLContext -> SDL.Texture -> IO ()
+loop event ctx textTexture = do
+    let loc = SDL.Rect 320 240 150 100
+    _ <- SDL.renderClear (getRenderer ctx)
```
Comment (forestbelton) [inline app/Main.hs]: instead of `_ <- foo`, you can just do `foo`

File: orb.cabal [hunk 3]
```diff
@@ -16,12 +16,14 @@ cabal-version:       >=1.10
 library
   hs-source-dirs:      src
   exposed-modules:     DOM
+                       Node
```
Comment (forestbelton) [inline orb.cabal]: shouldn't there be a comma here?

**Comment by forestbelton** [comment]
can you remove the `.DS_Store` files?

File: app/Main.hs [hunk 4]
```diff
@@ -7,7 +7,9 @@ import Foreign.C.Types
 import Foreign.C.String
 import Foreign.Marshal.Alloc
 import Foreign.Storable
+import Foreign (with)
```
Comment (forestbelton) [inline app/Main.hs]: hmm, can't find docs for this, mind giving a link?

File: app/Main.hs [hunk 5]
```diff
@@ -26,12 +28,35 @@ initScreen width height = do
...
 main = do
     ctx <- initScreen 800 600
     ptrEvent <- malloc
-    draw ctx
-    eventLoop ptrEvent ctx
+    --draw ctx
+    
+    TTF.withInit $ do
+        font <- TTF.openFont arial 150
+        textSurface <- TTF.renderUTF8Solid font "test" (SDL.Color 255 255 255 0)
```
Comment (forestbelton) [inline app/Main.hs]: wouldn't `rgba(255, 255, 255, 0` be completely transparent?

File: app/Main.hs [hunk 6]
```diff
@@ -26,12 +28,35 @@ initScreen width height = do
...
+        textTexture <- SDL.createTextureFromSurface (getRenderer ctx) textSurface
+        loop ptrEvent ctx textTexture
```
Comment (forestbelton) [inline app/Main.hs]: can you make a new constructor for `DisplayCommand` that takes a position, font name, font weight, color, and text, and basically encapsulates this logic? that way, the main application doesn't have to really know about the specifics of rendering, and it'll be easier to generate display text commands during the painting process

Comment (forestbelton) [inline app/Main.hs]: you'll want to add a new case [here](https://github.com/forestbelton/orb/blob/master/src/Paint.hs#L11), and then handle it in `paintCmd`.

**Comment by forestbelton** [comment]
i get this when trying to run `stack build`, any ideas?

```
Configuring sdl2-ttf-0.2.1...
    Building sdl2-ttf-0.2.1...
    Preprocessing library sdl2-ttf-0.2.1...
    Types.hsc:1:10: fatal error: 'SDL2/SDL_ttf.h' file not found
    #include "SDL2/SDL_ttf.h"
```

i did a `brew install sdl2_ttf` beforehand

**Comment by kyle-sammons** [comment]
If I recall correctly, sdl2-ttf wanted GHC 7.10.2. Have you tried running; "stack setup"?

File: app/Main.hs [hunk 7]
```diff
@@ -30,17 +30,25 @@ initScreen width height = do
...
+    SDL.quit
+    return ()
+
 main :: IO ()
 main = do
     ctx <- initScreen 800 600
     ptrEvent <- malloc
     draw ctx
     eventLoop ptrEvent ctx
+    cleanUp ctx
```
Comment (forestbelton) [inline app/Main.hs]: this should be called when we handle the quit event, not here

File: src/Paint.hs [hunk 8]
```diff
@@ -19,6 +25,16 @@ paintCmd re cmd = case cmd of
             SDL.setRenderDrawColor re r g b a
             SDL.renderFillRect re ptrRect
             return ()
+    FontData (x, y) name weight color text -> do
+         TTF.withInit $ do
+            font <- TTF.openFont arial 150
```
Comment (forestbelton) [inline src/Paint.hs]: shouldn't this be `name` ?

File: src/Paint.hs [hunk 9]
```diff
@@ -19,6 +25,16 @@ paintCmd re cmd = case cmd of
...
+            textSurface <- TTF.renderUTF8Solid font "test" (SDL.Color 255 255 255 1)
```
Comment (forestbelton) [inline src/Paint.hs]: and this `text`

File: src/Paint.hs [hunk 10]
```diff
@@ -19,6 +25,16 @@ paintCmd re cmd = case cmd of
             SDL.setRenderDrawColor re r g b a
             SDL.renderFillRect re ptrRect
             return ()
+    FontData (x, y) name weight color text -> do
+         TTF.withInit $ do
```
Comment (forestbelton) [inline src/Paint.hs]: `withInit` should live out in `Main`, like i mentioned

File: .gitignore [hunk 11]
```diff
@@ -1,2 +1,4 @@
 .stack-work/
 *.swp
+.gitignore
```
Comment (forestbelton) [inline .gitignore]: wat

**Comment by forestbelton** [comment]
:+1:

**Pull request merged**